CAF fails to compile on 32-bit Linux with GCC

The master branch can not be compiled in Raspbian. 
I use raspberry pi 3 and my GCC version is gcc version 8.3.0 (Raspbian 8.3.0-6+rpi1). It gave me these errors:

```
[ 35%] Linking CXX executable caf-core-test
/usr/bin/ld: CMakeFiles/caf-core-test.dir/test/actor_clock.cpp.o: in function `std::__atomic_base<long long>::operator++()':
actor_clock.cpp:(.text._ZNSt13__atomic_baseIxEppEv[_ZNSt13__atomic_baseIxEppEv]+0x28): undefined reference to `__atomic_fetch_add_8'
/usr/bin/ld: CMakeFiles/caf-core-test.dir/test/telemetry/collector/prometheus.cpp.o: in function `caf::telemetry::int_gauge::inc(long long)':
prometheus.cpp:(.text._ZN3caf9telemetry9int_gauge3incEx[_ZN3caf9telemetry9int_gauge3incEx]+0x3c): undefined reference to `__atomic_fetch_add_8'
/usr/bin/ld: CMakeFiles/caf-core-test.dir/test/telemetry/collector/prometheus.cpp.o: in function `caf::telemetry::int_gauge::value(long long)':
prometheus.cpp:(.text._ZN3caf9telemetry9int_gauge5valueEx[_ZN3caf9telemetry9int_gauge5valueEx]+0x4c): undefined reference to `__atomic_store_8'
/usr/bin/ld: CMakeFiles/caf-core-test.dir/test/telemetry/counter.cpp.o: in function `caf::telemetry::int_gauge::value() const':
counter.cpp:(.text._ZNK3caf9telemetry9int_gauge5valueEv[_ZNK3caf9telemetry9int_gauge5valueEv]+0x40): undefined reference to `__atomic_load_8'
/usr/bin/ld: CMakeFiles/caf-core-test.dir/test/telemetry/counter.cpp.o: in function `std::atomic<double>::load(std::memory_order) const':
counter.cpp:(.text._ZNKSt6atomicIdE4loadESt12memory_order[_ZNKSt6atomicIdE4loadESt12memory_order]+0x38): undefined reference to `__atomic_load_8'
/usr/bin/ld: CMakeFiles/caf-core-test.dir/test/telemetry/counter.cpp.o: in function `std::atomic<double>::compare_exchange_weak(double&, double, std::memory_order, std::memory_order)':
counter.cpp:(.text._ZNSt6atomicIdE21compare_exchange_weakERddSt12memory_orderS2_[_ZNSt6atomicIdE21compare_exchange_weakERddSt12memory_orderS2_]+0x78): undefined reference to `__atomic_compare_exchange_8'
/usr/bin/ld: CMakeFiles/caf-core-test.dir/test/telemetry/gauge.cpp.o: in function `caf::telemetry::int_gauge::dec(long long)':
gauge.cpp:(.text._ZN3caf9telemetry9int_gauge3decEx[_ZN3caf9telemetry9int_gauge3decEx]+0x3c): undefined reference to `__atomic_fetch_sub_8'
/usr/bin/ld: CMakeFiles/caf-core-test.dir/test/telemetry/gauge.cpp.o: in function `std::atomic<double>::store(double, std::memory_order)':
gauge.cpp:(.text._ZNSt6atomicIdE5storeEdSt12memory_order[_ZNSt6atomicIdE5storeEdSt12memory_order]+0x58): undefined reference to `__atomic_store_8'
/usr/bin/ld: CMakeFiles/caf-core-test.dir/test/telemetry/gauge.cpp.o: in function `std::__atomic_base<long long>::operator--()':
gauge.cpp:(.text._ZNSt13__atomic_baseIxEmmEv[_ZNSt13__atomic_baseIxEmmEv]+0x28): undefined reference to `__atomic_fetch_sub_8'
collect2: error: ld returned 1 exit status
```

But the actor framework 0.17.6 can be compiled in my Raspbian without any problem.

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -38,7 +38,7 @@ is based on [Keep a Changelog](https://keepachangelog.com).
 - Building executables that link to CAF on 32-bit Linux versions using GCC
   failed due to undefined references to `__atomic_fetch` symbols. Adding a CMake
   dependency for `caf_core` to libatomic gets executables to compile and link as
-  expected.
+  expected (#1153).
 
 ## [0.18.0-rc.1] - 2020-09-09
 

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -35,6 +35,10 @@ is based on [Keep a Changelog](https://keepachangelog.com).
   case or do unexpected things in the latter case. The stringification inspector
   now matches precisely on pointer types to stop the compiler from doing
   implicit conversions in the first place.
+- Building executables that link to CAF on 32-bit Linux versions using GCC
+  failed due to undefined references to `__atomic_fetch` symbols. Adding a CMake
+  dependency for `caf_core` to libatomic gets executables to compile and link as
+  expected.
 
 ## [0.18.0-rc.1] - 2020-09-09
 

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -5,6 +5,7 @@ project(CAF CXX)
 
 include(CMakeDependentOption)      # Conditional default values
 include(CMakePackageConfigHelpers) # For creating .cmake files
+include(CheckCXXSourceCompiles)    # For checking whether some C++ features work
 include(GNUInstallDirs)            # Sets default install paths
 include(GenerateExportHeader)      # Auto-generates dllexport macros
 

--- a/libcaf_core/CMakeLists.txt
+++ b/libcaf_core/CMakeLists.txt
@@ -194,7 +194,26 @@ caf_core_set_default_properties(libcaf_core_obj libcaf_core)
 
 caf_export_and_install_lib(core)
 install(FILES "${CMAKE_BINARY_DIR}/caf/detail/build_config.hpp"
-          DESTINATION "${CMAKE_INSTALL_INCLUDEDIR}/caf/detail")
+        DESTINATION "${CMAKE_INSTALL_INCLUDEDIR}/caf/detail")
+
+# Check whether we need to link against libatomic.
+if(NOT CMAKE_CROSSCOMPILING)
+  set(snippet "#include <cstdint>
+               #include <atomic>
+               std::atomic<uint64_t> x;
+               int main(int, char**) { return static_cast<int>(x.load()); }")
+  check_cxx_source_compiles("${snippet}" has_64bit_atomic)
+  if(NOT has_64bit_atomic)
+    set(required_libs_backup "${CMAKE_REQUIRED_LIBRARIES}")
+    list(APPEND CMAKE_REQUIRED_LIBRARIES "atomic")
+    check_cxx_source_compiles("${snippet}" has_64bit_atomic_with_libatomic)
+    set(CMAKE_REQUIRED_LIBRARIES "${required_libs_backup}")
+    if(NOT has_64bit_atomic_with_libatomic)
+      message(FATAL_ERROR "Unable to compile code with std::atomic<uint64_t>")
+    endif()
+    target_link_libraries(libcaf_core INTERFACE atomic)
+  endif()
+endif()
 
 # -- build unit tests ----------------------------------------------------------
 

